homeassistant/components/androidtv/__init__.py

```diff
@@ -1 +1,232 @@
+    CONF_ADBKEY,
+    CONF_STATE_DETECTION_RULES,
+    DEFAULT_ADB_SERVER_PORT,
+    DEVICE_ANDROIDTV,
+    DEVICE_FIRETV,
+    DOMAIN,
+    MIGRATION_DATA,
+    PROP_SERIALNO,
+    SIGNAL_CONFIG_ENTITY,
+)
+
+PLATFORMS = [MP_DOMAIN]
+RELOAD_OPTIONS = [CONF_STATE_DETECTION_RULES]
+
+_LOGGER = logging.getLogger(__name__)
+
+
+def _setup_androidtv(hass, config):
+    """Generate an ADB key (if needed) and load it."""
+    adbkey = config.get(CONF_ADBKEY, hass.config.path(STORAGE_DIR, "androidtv_adbkey"))
+    if CONF_ADB_SERVER_IP not in config:
+        # Use "adb_shell" (Python ADB implementation)
+        if not os.path.isfile(adbkey):
+            # Generate ADB key files
+            keygen(adbkey)
+
+        # Load the ADB key
+        signer = ADBPythonSync.load_adbkey(adbkey)
+        adb_log = f"using Python ADB implementation with adbkey='{adbkey}'"
+
+    else:
+        # Use "pure-python-adb" (communicate with ADB server)
+        signer = None
+        adb_log = f"using ADB server at {config[CONF_ADB_SERVER_IP]}:{config[CONF_ADB_SERVER_PORT]}"
+
+    return adbkey, signer, adb_log
+
+
+def validate_state_det_rules(state_det_rules):
+    """Validate a string that contain state detection rules and return a dict."""
+    if not state_det_rules:
+        return None
+
+    try:
+        json_rules = json.loads(state_det_rules)
+    except ValueError:
+        _LOGGER.warning("Error loading state detection rules")
+        return None
+    try:
+        state_detection_rules_validator(json_rules, ValueError)
+    except ValueError as exc:
+        _LOGGER.warning("Invalid state detection rules: %s", exc)
+        return None
+    return json_rules
+
+
+async def async_connect_androidtv(
+    hass, config, *, state_detection_rules=None, timeout=30.0
+):
+    """Connect to Android device."""
+    address = f"{config[CONF_HOST]}:{config[CONF_PORT]}"
+
+    adbkey, signer, adb_log = await hass.async_add_executor_job(
+        _setup_androidtv, hass, config
+    )
+
+    aftv = await setup(
+        config[CONF_HOST],
+        config[CONF_PORT],
+        adbkey,
+        config.get(CONF_ADB_SERVER_IP),
+        config.get(CONF_ADB_SERVER_PORT, DEFAULT_ADB_SERVER_PORT),
+        state_detection_rules,
+        config[CONF_DEVICE_CLASS],
+        timeout,
+        signer,
+    )
+
+    if not aftv.available:
+        # Determine the name that will be used for the device in the log
+        if config[CONF_DEVICE_CLASS] == DEVICE_ANDROIDTV:
+            device_name = "Android TV device"
+        elif config[CONF_DEVICE_CLASS] == DEVICE_FIRETV:
+            device_name = "Fire TV device"
+        else:
+            device_name = "Android TV / Fire TV device"
+
+        _LOGGER.warning(
+            "Could not connect to %s at %s %s", device_name, address, adb_log
+        )
+        return None
+
+    if CONF_ADB_SERVER_IP in config:
+        _LOGGER.warning(
+            "Adb server is deprecated. Please consider connecting directly to your TV"
```

Comment:
The ADB add-on is deprecated, but using an ADB server in the `androidtv` library is not -- I have no plans to remove that functionality. 

If a user wants to maintain their own ADB server, that's perfectly fine. But if the Python ADB implementation (`adb-shell`) works for that user then I don't know why they would go to that trouble. An ADB server also makes debugging much more challenging. 

I'll leave it up to you to decide if you want to keep, reword, or remove this log statement.

Reply:
In this case I think that log should be removed or at least moved to `info` level.

Reply:
I suggest rewording it to state that the ADB add-on is deprecated, not the ADB server config option.

If you really want to get fancy, I believe there's a way to check if an add-on is installed, so you could only log this if the ADB add-on is installed. But I'm sure the vast majority of people using an ADB server are utilizing the add-on, so it might be more trouble than it's worth to do that.

Reply:
Perhaps we should remove it then. I can't easily get the addon installed for testing and I'm note sure what the exact addon slug is called for `async_get_addon_info`.

Reply:
I removed this warning. I suppose that is the obsolete plugin that should raise it.